Use latest version of cppcheck (2.15.0) to analyze codebase

eduar-hte opened the pull request:
## what

 * Update the `cppcheck` GitHub workflow to run on the latest version of that tool (2.15.0) for better codebase analysis (more aligned with that of Sonarcloud).
   * NOTE: This PR references `cppcheck` version 2.14.2 because that was the last version at the time it was created. However, the GitHub macOS runner will now install 2.15.0 (the latest version), which has been released since then (Aug 31, 2024).
 * Update the `cppcheck` configuration to have the analysis be performed using C++17, which is aligned with the build configuration for the project.
 * Address warnings identified by the newer version of the tool.
   * This should also address some similar Sonarcloud issues as well.

## why

The latest version of `cppcheck` includes improved analysis and address some limitations of the version currently used (2.7.1, from Feb, 2022).

For example, `std::string_view` instances are expected to be passed by value (see https://quuxplusone.github.io/blog/2021/11/09/pass-string-view-by-value/), but the older version of `cppcheck` would report a warning incorrectly stating that the argument should be passed by reference instead. This was addressed in danmar/cppcheck#3817

## notes

The `cppcheck` GitHub workflow now runs on the macOS image (instead of the Ubuntu/Linux image previously used) for two reasons:

 * The Ubuntu/Linux image didn't have the latest version of `cppcheck` available, so we needed to manually checkout the `cppcheck` repository and compile the tool, which would make the workflow more complex and take more time.
 * The `cppcheck` build in Ubuntu/Linux would run much slower than the package available on macOS (probably due to some additional build instructions being required to improve performance which I couldn't determine/find).

eduar-hte committed:
Use latest version of cppcheck (2.14.2)
- Run cppcheck on MacOS to use a newer version of cppcheck

eduar-hte committed:
Added const reported by cppcheck 2.14

eduar-hte committed:
Ignore cppcheck warnings: normalCheckLevelMaxBranches (Limiting analysis of branches. Use --check-level=exhaustive to analyze all branches.)

eduar-hte committed:
Added missing override keyword as reported by cppcheck 2.14

eduar-hte committed:
Address cppcheck warnings: uselessOverride (The function '...' overrides a function in a base class but is identical to the overridden function)

eduar-hte committed:
Suppress warnings on seclang-parser.hh
warning: seclang-parser.hh,2116,warning,duplInheritedMember,The struct 'basic_symbol < by_kind >' defines member function with name 'clear' also defined in its parent struct 'by_kind'.
warning: seclang-parser.hh,2376,warning,duplInheritedMember,The struct 'basic_symbol < by_kind >' defines member function with name 'type_get' also defined in its parent struct 'by_kind'.
warning: seclang-parser.hh,2116,warning,duplInheritedMember,The struct 'basic_symbol < by_state >' defines member function with name 'clear' also defined in its parent struct 'by_state'.
warning: seclang-parser.hh,2120,style,constVariableReference,Variable 'yysym' can be declared as reference to const

eduar-hte committed:
Updated cppcheck config
- Do not scan third-party libraries (others dir)
- Use standard C++17 for checks (defaults to C++20)

eduar-hte committed:
Add cppcheck suppressions for false positives

eduar-hte committed:
Address cppcheck warnings generated after addressing Sonarcloud suggestions
- The following two warnings were generated after introducing the change
  to instantiate the DigestImpl template with the address of mbedtls_md5
  or mbedtls_sha1:
  - warning: src/utils/sha1.h,62,error,danglingTemporaryLifetime,Using
    pointer that is a temporary.
  - warning: src/utils/sha1.h,60,style,constVariablePointer,Variable
    'ret' can be declared as pointer to const
- See https://github.com/owasp-modsecurity/ModSecurity/pull/3231#issuecomment-2312511500

eduar-hte committed:
Refactor to remove duplicate code in ValidateSchema & ValidateDTD
- Reported by Sonarcloud

eduar-hte committed:
Avoid this unnecessary copy by using a "const" reference.
- Reported by Sonarcloud

eduar-hte commented:
> Failed conditions ![](https://camo.githubusercontent.com/bd1d5234b37410841ad8cd9d85c84eeecaec92ae18a95c926ae3a59a17c29a65/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636f6d6d6f6e2f6661696c65642d313670782e706e67) [B Maintainability Rating on New Code](https://sonarcloud.io/dashboard?id=owasp-modsecurity_ModSecurity&pullRequest=3283) (required ≥ A)

I added two commits (unrelated to this PR) to address Sonarcloud reports. The first blocked analysis of the commits due to Sonarcloud finding (existing) duplicate code in `ValidateSchema` & `ValidateDTD` where initially there was a single line change to include the `override` keyword in those classes' destructors (and then the destructor was removed altogether).

The set of "new" (between quotes) issues reported by Sonarcloud after these changes are:

 *  `src/actions/exec.h`
   *  `Add a using-declaration to this derived class to inherit the constructors of "Action", and remove the ones you manually duplicated. Note that this may add other constructors to your derived class.` 
     * Addressing this issue would include additional constructors to `Exec`, which are not currently supported, so it's not an equivalent change.
 * `src/operators/validate_dtd.h` & `src/operators/validate_schema.h`
   * `Remove use of ellipsis notation` (2)
   * `Replace this use of "void *" with a more meaningful type.` (6)
   * `Make the type of this parameter a pointer-to-const. The current type of this unnamed parameter is "void *"` (3)
   * `Replace "reinterpret_cast" with a "static_cast"` (2)
     * All these are required by the signature of `libxml2` C-style callback functions.
 * `test/regression/regression.cc`
   * `Refactor this function to reduce its Cognitive Complexity from ...` (5)
     * Existing code, unrelated to these changes.

airween commented on src/actions/transformations/parity_zero_7bit.cc:
```diff
@@ -22,7 +22,7 @@ namespace modsecurity::actions::transformations {
     if (value.empty()) return false;
 
-    for(auto &c : value) {
+    for(auto &c : value) { // cppcheck-suppress constVariableReference ; false positive
```
I wonder why `cppcheck` hasn't indicated this until now? It's because  of the new version or that?

eduar-hte replied:
Yes, I understand that while they may add new features and address bugs in newer versions, some of these changes may have introduced inaccurate reports as well. 🤷‍♂️

airween commented on src/compat/msvc.h:
```diff
@@ -14,6 +14,7 @@
 #define pclose       _pclose
 
 inline tm* localtime_r(const time_t* tin, tm* tout) {
+  // cppcheck-suppress[uninitvar, ctuuninitvar]
```
Same as above.

eduar-hte replied:
Same as above. 😊

airween replied:
Thanks! :smile:

airween commented:
Hi @eduar-hte,

many thanks for this great PR. Also thanks for adding `false positive` labels to `cppcheck`  suppression's.

I have only one question regarding the new inline suppression's: are those there because of the new `cppcheck`? Why those haven't appeared until now? (I marked the first two items and later I saw that there are many others there)

About SonarCloud: I already sent you an invitation to have an admin role on SonarCloud - haven't you received that? With that you could mark those issues (that you explain every time really accurately) as FP and suppress them.

eduar-hte committed:
Use initialization list to initialize m_service
- This is correct because base class is initialized before members are
  initialized.
- Removes cppcheck suppression by addressing reported issue.
- Leverage C++11's 'default member initializer' to initialize m_provider
  & m_demandsPassword and address Sonarcloud issue.

eduar-hte committed:
Remove cppcheck suppression by replacing use of local variable to alias this->m_variables
- The name of the local variable would clash with the namespace of the
  same name, which may have lead cppcheck to think the variable was not
  used.

eduar-hte committed:
Replace usage of sscanf with strtol to remove cppcheck inline suppression

eduar-hte committed:
Make GeoLookup::debug function static (and non-member), as suggested by cppcheck.

eduar-hte committed:
Remove no longer needed cppcheck inline suppressions.

eduar-hte commented:
> I have only one question regarding the new inline suppression's: are those there because of the new `cppcheck`? Why those haven't appeared until now? (I marked the first two items and later I saw that there are many others there)

Yes, the new version reported some new issues, some of which were addressed with changes to the code as they were correct (see commits related to using the `override` keyword or adding `const` for example).

At the same time, the new version also introduced a (limited) number of inaccurate issues, which I had to address by adding the inline suppressions you noticed (those are grouped in commit d053ec6).

I understand that while they may add new features and address bugs in newer versions, some of these changes may have introduced inaccurate reports as well. 🤷‍♂️

I've spent some time today on this PR again to do an experiment and remove all current inline suppressions and check which ones are still reported, because it may be the case that they're no longer necessary. As you can see in a0490bd, I could do away with a few of them, which is nice.

Additionally, when reviewing every one that was still reported by `cppcheck` I realized I could implement the suggestion reported by the tool instead of suppressing it, so a few more of them could be removed from the codebase.

> About SonarCloud: I already sent you an invitation to have an admin role on SonarCloud - haven't you received that? With that you could mark those issues (that you explain every time really accurately) as FP and suppress them.

Sorry, I didn't see that. In any case, I'm not sure whether we should remove some of the issues reported by Sonarcloud that I detail in some of the PRs. My goal with that is to document which issues are unrelated to changes in the PR (they currently exist in the codebase and Sonarcloud reports them in the PR because 
some code has been changed around the existing issue).

I would group the issues in those lists in at least three groups:

 * There are some that are false positives like the ones related to the libxml2 callback functions.
 * There are some that may be sensible, but addressing them would be unrelated to the current PR, like the ones about code complexity (which could also lead to issues by trying to address an issue on stable code that has not changed in years, so the tradeoff for addressing an issue there is tricky). 
 * Then there are some that are opinionated, like [this one](https://sonarcloud.io/project/issues?cleanCodeAttributeCategories=INTENTIONAL&sinceLeakPeriod=true&issueStatuses=OPEN%2CCONFIRMED&pullRequest=3283&id=owasp-modsecurity_ModSecurity&open=AZKwjxkjPqSupH1YTEAv) that just came up after commit 651536f. I think that c++11 default member initializers are nice, but a matter of style whether to use them or not.

About the first group, I think it would make sense to flag them, but I think it'd be better to do so through a mechanism similar to cppcheck's inline suppressions. I tried to look for that, but I couldn't find it ([this Sonarcloud community issue](https://community.sonarsource.com/t/suppress-issue-in-c-source-file/43154) seems to be about that).

airween commented:
> Yes, the new version reported some new issues, some of which were addressed with changes to the code as they were correct (see commits related to using the `override` keyword or adding `const` for example).

I saw and I though you aligned those because of this reason.
 
> At the same time, the new version also introduced a (limited) number of inaccurate issues, which I had to address by adding the inline suppressions you noticed (those are grouped in commit [d053ec6](https://github.com/owasp-modsecurity/ModSecurity/commit/d053ec6de6a9a1b14dbc453fef7f12efcd202b90)).

Thanks,

> I understand that while they may add new features and address bugs in newer versions, some of these changes may have introduced inaccurate reports as well. 🤷‍♂️

Thanks again.

I saw you added some more stuff (removed some inline suppression). Let me know if you finish this PR.

eduar-hte commented:
> I saw you added some more stuff (removed some inline suppression). Let me know if you finish this PR.

That's right, I did away with a few of the cppcheck inline suppressions, as I ended up explaining in an edit of my previous reply (which I was in the process of writing when the build with the new changes completed so I decided to merge them into the PR and then resume the lengthy reply 😁).

PS: I ended up even addressing the new issue Sonarcloud reported and that I mentioned here:

> Then there are some that are opinionated, like [this one](https://sonarcloud.io/project/issues?cleanCodeAttributeCategories=INTENTIONAL&sinceLeakPeriod=true&issueStatuses=OPEN%2CCONFIRMED&pullRequest=3283&id=owasp-modsecurity_ModSecurity&open=AZKwjxkjPqSupH1YTEAv) that just came up after commit https://github.com/owasp-modsecurity/ModSecurity/commit/651536fa0eacd60ca5b99ca0cbe675f0674e327b. I think that c++11 default member initializers are nice, but a matter of style whether to use them or not.

just not to add up to the number of outstanding Sonarcloud issues.

I'm now done with this PR.

airween reviewed (approved):
LGTM

airween commented:
Thanks for the PR again and for the detailed explanations.

I'm merging this now.
